### DISCUSSION TITLE
Whole-flash-erase instead of block-by-block erase before programming?

### DISCUSSION THREAD

**Discussion by obra** [stlink-org/stlink#1238] [Q&A (resolved)] [answered]
We're using st-flash as part of our in-factory toolchain for programming the microcontrollers in our product.  Right now, the slowest part of the process, by far, is the block-by-block erase process just before flashing. 

A whole-device erase with `st-flash erase`, is, of course, lightning fast.

It feels like what I want would be an option to force whole-device erase. rather than block-by-block erase during `st-flash write` 

Is this something that ought to exist but just hasn't happened yet, is there a better way to do what I want, or is what I want just wrong for some reason that I'm not quite seeing?

Thanks!

**Comment by Nightwalker-87** [comment]
@obra Let me review this in correlation to the current implementation in the code base.

**Comment by Nightwalker-87** [comment]
@obra Actually, I don't see any general technical reason why this shouldn't be possible.
What I can't tell is, whether any specific way of implementation is opposing such an attempt.

**Comment by Nightwalker-87** [comment]
@Ant-ON How do you think about this?

**Comment by Ant-ON** [comment]
@Nightwalker-87 Typically, flashing tools provide the option to full flash memory erase or only used memory pages

**Reply by Nightwalker-87** [reply]
The stlink toolset seems to as well (never made use of it before though):
`st-flash [--debug] [--connect-under-reset] [--hot-plug] [--freq=<kHz>] [--serial <serial>] erase [addr] [size]`
with `size` needing to be aligned to page sizes. `addr` does not necessarily need to point to the start address, but as far as I'm aware of needs to be a multiple of the predefined page size as well. However the message `Mass erase completed successfully.` does appear in each case ...

**Comment by Nightwalker-87** [comment, marked as answer]
Feature added in #1397.
